pdns/packethandler.cc

```diff
@@ -456,7 +456,7 @@ int PacketHandler::doAdditionalProcessingAndDropAA(DNSPacket& p, std::unique_ptr
          !(i->dr.d_type==QType::MX || i->dr.d_type==QType::NS || i->dr.d_type==QType::SRV))
         continue;
 
-      if(r->d.aa && i->dr.d_name.countLabels() && i->dr.d_type==QType::NS && !B.getSOA(i->dr.d_name,sd) && !retargeted) { // drop AA in case of non-SOA-level NS answer, except for root referral
+      if(r->d.aa && i->dr.d_name.countLabels() && i->dr.d_type==QType::NS && !i->auth && !retargeted) { // drop AA in case of non-SOA-level NS answer, except for root referral
```

Comment:
Is `auth` always correctly set, even in non DNSSEC zones or non DNSSEC-capable backends?

Reply:
This is indeed a potential problem - one where I'm not sure how to test this / what would actually break.

IMO we could also call getAuth instead, which would at least be correct over getSOA. The perf gain from avoiding another DB lookup could be done in a separate PR.

Reply:
In #9478 @mind04 fixed this bit.